lib/xgu/xgu.h

```diff
@@ -0,0 +1,969 @@
+    assert(texture_index < XGU_TEXTURE_COUNT);
+    return push_command_parameter(p, NV097_SET_TEXTURE_OFFSET + texture_index*64, (uint32_t)offset);
+}
+
+inline
+uint32_t* xgu_set_texture_format(uint32_t* p, unsigned int texture_index, uint8_t context_dma, bool cubemap_enable,
+                                 XguBorderSrc border_src, uint8_t dimensionality,
+                                 XguTexFormatColor format, uint8_t mipmap_levels,
+                                 uint8_t u_size, uint8_t v_size, uint8_t p_size) {
+    assert(texture_index < XGU_TEXTURE_COUNT);
+    return push_command_parameter(p, NV097_SET_TEXTURE_FORMAT + texture_index*64,
+                                  XGU_MASK(NV097_SET_TEXTURE_FORMAT_CONTEXT_DMA, context_dma) |
+                                  XGU_MASK(NV097_SET_TEXTURE_FORMAT_CUBEMAP_ENABLE, cubemap_enable) |
+                                  XGU_MASK(NV097_SET_TEXTURE_FORMAT_BORDER_SOURCE, border_src) |
+                                  XGU_MASK(NV097_SET_TEXTURE_FORMAT_DIMENSIONALITY, dimensionality) |
+                                  XGU_MASK(NV097_SET_TEXTURE_FORMAT_COLOR, format) |
+                                  XGU_MASK(NV097_SET_TEXTURE_FORMAT_MIPMAP_LEVELS, mipmap_levels) |
+                                  XGU_MASK(NV097_SET_TEXTURE_FORMAT_BASE_SIZE_U, u_size) |
+                                  XGU_MASK(NV097_SET_TEXTURE_FORMAT_BASE_SIZE_V, v_size) |
+                                  XGU_MASK(NV097_SET_TEXTURE_FORMAT_BASE_SIZE_P, p_size));
+}
+
+inline
+uint32_t* xgu_set_texture_address(uint32_t* p, unsigned int texture_index, uint32_t address) {
+    assert(texture_index < XGU_TEXTURE_COUNT);
+    return push_command_parameter(p, NV097_SET_TEXTURE_ADDRESS + texture_index*64, address);
+}
+
+inline
+uint32_t* xgu_set_texture_control0(uint32_t* p, unsigned int texture_index, bool enable, uint16_t min_lod, uint16_t max_lod) {
+    assert(texture_index < XGU_TEXTURE_COUNT);
+    return push_command_parameter(p, NV097_SET_TEXTURE_CONTROL0 + texture_index*64,
+                                  (enable ? NV097_SET_TEXTURE_CONTROL0_ENABLE : 0) |
+                                  XGU_MASK(NV097_SET_TEXTURE_CONTROL0_MIN_LOD_CLAMP, min_lod) |
+                                  XGU_MASK(NV097_SET_TEXTURE_CONTROL0_MAX_LOD_CLAMP, max_lod));
+}
+
+inline
+uint32_t* xgu_set_texture_control1(uint32_t* p, unsigned int texture_index, uint16_t pitch) {
+    assert(texture_index < XGU_TEXTURE_COUNT);
+    return push_command_parameter(p, NV097_SET_TEXTURE_CONTROL1 + texture_index*64,
+                                  XGU_MASK(NV097_SET_TEXTURE_CONTROL1_IMAGE_PITCH, pitch));
+}
+
+inline
+uint32_t* xgu_set_texture_filter(uint32_t* p, unsigned int texture_index, uint16_t lod_bias, uint8_t filter_min, uint8_t filter_mag,
+                                 bool r_signed, bool g_signed, bool b_signed, bool a_signed) {
```

Comment:
*(This quotes @Voxel9 so the top level comment can be removed; Thanks very much for this detailed report!)*

> `xgu_set_texture_filter` doesn't work on real hardware. I think it's expecting bits 13-15 to be set to something not 0? There's no bitmask in nv2a_regs.h associated with this field, neither is it set by `xgu_set_texture_filter`.

That field is actually the convolution filter.

> The pbkit mesh sample sets this field with a value of 2, which works on both hardware and XQEMU. A value of 1 also works. But it seems `xgu_set_texture_filter` always sets it to 0, causing an invalid data error on hardware.

It can be 1 (Quincunx) or 2 (Gaussian). It should only be used if the min_filter is 7 (convolution), or the mag_filter is 4 (convolution). It should be unused for all other modes.

There are 2 options to handle this API wise:

**Make it part of min-filter and mag-filter**

- Make an enum for the Min filters (Including ConvolutionGuassian and ConvolutionQuincunx)
- Make an enum for the Mag filters (Including ConvolutionGuassian and ConvolutionQuincunx)
- If both of them are convolution, assert that they are the same convolution type (2x Gaussian or 2x Quincunx).

Less confusing as people will not have a switch which does nothing in many situations.
But also more confusing because it's unexpected that both have to have the same convolution type.
Also a disadvantage that it could not be controlled with the other modes, so debugging for XQEMU would be trickier, as it's hard to set a different min/mag filter without also forcing a certain convolution, even then only certain convolution would be exposed and playing with it expects that the user knows the internal bitfields.

**Make it a separate option**

Have it be an independent parameter, that is often unused.

Can be confusing as it's an option which rarely does anything.
Advantage is that it's always possible to force a mode, so it's easier to verify the assumption if it does anything even when the min/mag filters aren't convolution. Also allows to set other values than 1 and 2 for testing.

---

Eitherway, min/mag filter should be 2 enum types. And the convolution filter would have to be another one, or part of the min/mag filter values.

---

*(@Voxel9 in the future, if you know a specific function is broken, add a code comment (like this) on the affected code lines. That way, GitHub allows a more threaded discussion and it also allows hiding of the entire conversation)*